microcloud/cmd/microcloud/main_init.go

```diff
@@ -414,6 +419,82 @@ func AddPeers(sh *service.Handler, systems map[string]InitSystem) error {
 	return nil
 }
 
+func ensureSystemsOutsideGateway(gateway string, localAddr string, systems map[string]InitSystem) error {
+	_, ip4Net, err := net.ParseCIDR(gateway)
+	if err != nil {
+		return fmt.Errorf("Invalid ipv4.gateway for UPLINK: %w", err)
+	}
+
+	localAddrIP := net.ParseIP(localAddr)
+	if localAddrIP == nil {
+		return fmt.Errorf("Invalid local address %q", localAddr)
+	}
+
+	if ip4Net.Contains(localAddrIP) {
+		return fmt.Errorf("UPLINK ipv4.gateway must not include local address %q", localAddr)
+	}
+
+	for _, system := range systems {
+		systemAddr := net.ParseIP(system.ServerInfo.Address)
+		if systemAddr == nil {
+			return fmt.Errorf("Invalid address %q for system %q", system.ServerInfo.Address, system.ServerInfo.Name)
+		}
+
+		if ip4Net.Contains(systemAddr) {
+			return fmt.Errorf("UPLINK ipv4.gateway must not include system address %q", systemAddr)
+		}
+	}
+
+	return nil
+}
+
+func validateSystems(s *service.Handler, systems map[string]InitSystem) error {
+	_, bootstrap := systems[s.Name]
+	if !bootstrap {
+		return nil
+	}
+
+	for _, curSystem := range systems {
```

Comment:
Looks like you call `ensureSystemsOutsideGateway` twice for each qualifying network, and that is repeated for each system. Actually, the OVN networks should be identical on all systems, so we can factor away the outer loop, by only considering the `Networks` field on one `InitSystem`.

How about iterating over the systems first, collecting a list of parsed IPs for each system address, and then only iterating over `systems[s.Name].Networks`. Then we can change `ensureSystemsOutsideGateway` to have a signature like
`func ensureSystemsOutsideGateway(gateway string, systemIPs []net.IP) error`. 

That way we won't be redundantly parsing the system addresses and we won't validate the same uplink network `len(systems)` times. What do you think?

Reply:
I've restructured the code around the assumption that all the UPLINK nets are the same; I think the result is more readable and the tests make better sense. Let me know what you think.